### DISCUSSION TITLE
Can this software create 3D meshes from a MIDI file?

### DISCUSSION THREAD

**Discussion by sebarious70** [JacquesLucke/animation_nodes#1962] [Q&A] [answered]
I'm trying to create a 3D scene from a midi file. The scene would need to have each note as a cube rectangle, and the translation from the music world into the 3D scene would be basic: 1 bar = 1 meter. So if you have a MIDI note that is 1 bar long, the long side of the rectangular cube would be 1 meter long. The other two dimensions would be 25 cm each. So, something like this:

<img width="872" alt="Screenshot 2024-06-28 at 23 08 13" src="https://github.com/JacquesLucke/animation_nodes/assets/157246475/f72d5e04-f4cb-431f-a4f3-932648d22d51">

I'm on the verge when it comes to the other two dimensions for shorter notes, because if I keep 25 cm for all of them, when I do notes that are 1/2, 1/4 and 1/8, it would look like this:

<img width="1019" alt="Screenshot 2024-06-28 at 23 13 39" src="https://github.com/JacquesLucke/animation_nodes/assets/157246475/46b0fcc3-be4a-4ae1-bf18-9fe73affc5e7">

But if I also shorten X and Z more or less proportionally, I think perhaps it would look better:

<img width="990" alt="Screenshot 2024-06-28 at 23 15 03" src="https://github.com/JacquesLucke/animation_nodes/assets/157246475/fa428480-7661-4b28-acb1-e0a0041225c1">

Regardless, my goal is to get something like this:

![Screenshot 2024-06-24 at 21 49 52](https://github.com/JacquesLucke/animation_nodes/assets/157246475/7ccddb3e-352f-4a5f-b9b0-53a8364f1b1b)

into something like this:

![Screenshot 2024-06-24 at 21 50 42](https://github.com/JacquesLucke/animation_nodes/assets/157246475/18f2a279-4157-4208-8230-24dc035fcc11)

Obviously without the flat notes, those are just to show the difference between the flat ones and how the 3D ones should be)

My secretary (that would be the free version of Chat GPT) pointed me to this excellent software when I asked it if there was any program that can generate 3D meshes from MIDI notes. So I started reading the documentation and I found some hope, but first I want to ask here if what I'm trying to achieve is possible with this add-on. I don't even need animation, just static notes, all laid out in a 3D scene.

Now, there's a section of the documentation that I spent a lot of time trying to build exactly as it shows, but I didn't get anywhere near the animation that shows there, or saw any meshes being built: [](https://docs.animation-nodes.com/tutorials/midi/#raw-note-information). The only thing I achieved was to move the first cube I created in -Y.

Now, I copied the node tree under "Raw Note Information" as best as I could, but there is one node, the group input all the way to the left that has two inputs, one is a MIDI track, and the other is Width. Well, "Width" is not a name that shows up when I click on the + and type width. In fact, it doesn't appear in anything. Either the documentation is outdated or Blender 4.1.0 has something so it doesn't show? I couldn't say. But other than that, I think all the rest are just like the example in the link above:

<img width="1822" alt="Screenshot 2024-06-28 at 23 39 15" src="https://github.com/JacquesLucke/animation_nodes/assets/157246475/a97c96f6-9de0-432e-89fd-fe8e778c8c3d">

So that's all. Basically I want to know if this add-on can be setup in a way that will ingest a MIDI file (and it won't be a super complicated one, just one track), and turn that into a 3D scene, and if so, if this is more or less the node setup I need, or if I need to go another route.

**Comment by OmarEmaraDev** [comment]
The Width input is just a float input whose name is Width. You can definitely create what you want using this software, I can provide an example file if you couldn't get it to work eventually.

**Reply by sebarious70** [reply]
Good, at least now I know it's possible. Now, what's the learning curve? Because what I need to do can be done manually by setting the viewport to top view and putting a long wide screenshot of the section of the song that I want to do this for as a background, then grab the 1 bar note, duplicate several times, then the same thing with the other lengths, and finally placing them to match the notes. Time consuming, but not terribly so.

On the other hand, if the automation nodes way means spending days learning the stuff, that's time that I don't have. Would love to, because it looks like a great thing to learn, but I don't have the luxury of time. 

Basically the thing I study music, and I have  some songs that I would like to upload to YouTube, but with visuals, and I have a waveform that I did in Fusion with a 3rd party plugin, but then I thought about adding a way to visualize parts of the song with a MIDI track in 3D. But this is something I want to upload as soon as possible, so I can't spend a week or two learning this. 

How difficult is it to get it to create the meshes from a simple MIDI track, only index 0? I've been trying some things this afternoon but I couldn't find the way to actually generate meshes from the MIDI track.

**Comment by OmarEmaraDev** [comment, marked as answer]
@sebarious70 I am not not sure about the learning curve, because it likely depends on how familiar are you with Blender, nodes, maths, and so on. But this particular thing you want to do is rather easy, so it is probably worth learning a bit, and we can help you along the process.

I made you a simple example to get you started:

![20240630-115910](https://github.com/JacquesLucke/animation_nodes/assets/25300994/65854236-c80d-436f-b10c-917166c98c51)

Just change the path to your MIDI file and you should get MIDI notes as meshes. The Get List Element retrieves the first track from the file. Then we get the note info of that track, and compute the note length by subtracting the on time from the off time.  The X position of the note is its On time, its Y position is its note number. The X scale of the note is the note length, while its Y and Z scales are fixes, which you may change based on some note info. 

[midiToMesh.zip](https://github.com/user-attachments/files/16043803/midiToMesh.zip)

**Reply by sebarious70** [reply]
Wow! Man, that's freaking awesome, thank you so much. I think I understand the node tree to some extent, but it would've taken me several days to get there. If I understand the flow correctly, I think it doesn't actually create a mesh, but you feed it the default cube and it takes it and duplicates it by as many notes as there are in the MIDI file, correct?

But yeah, really cool stuff. The notes are not 1 bar = 1 meter but it seems to me that they look better this way, so I won't touch it.

Thanks so much!!!

**Reply by OmarEmaraDev** [reply]
Yes. We load the default cube and duplicate it by the number of notes, each with a specific transformation.

**Reply by sebarious70** [reply]
I just a had an idea, but I don't know if it's possible. With the node setup you showed me, the script grabs the default cube, and duplicates it as many times as notes in the MIDI file. So one question I have is, is it possible to create one object per note?

But most importantly, here's what I would like to achieve. Right now I have a simple setup with two materials and a wireframe modifier to have the notes appear as cubes with glowing edges. Now, what I would like to accomplish is that when each note is playing, that specific cube that represents that note, change to a non-wireframe cube on which the emission shows from all its sides, not just the edges.

Is there a way to accomplish that? I know it sounds super complicated, and I don't have the first clue about where to even begin. I've been squeezing my brain for hours, trying out different things, looking for it online, but nothing.

**Reply by OmarEmaraDev** [reply]
@sebarious70 You can use an Object Instancer node to create an object for each note.
![20240704-201302](https://github.com/JacquesLucke/animation_nodes/assets/25300994/f06fdfbc-3ecd-40aa-8170-c4b9be8c86c4)

Maybe look into the Copy Modifiers and Object Material Output nodes to achieved what you want.

**Reply by sebarious70** [reply]
That's awesome! I applied this to the node flow you gave me previously and it works like a charm. 

<img width="2221" alt="Screenshot 2024-07-04 at 13 59 12" src="https://github.com/JacquesLucke/animation_nodes/assets/157246475/74bb8015-17ff-4ab9-a7c9-649e0be769bd">

One thing I noticed is that it transfers the material slots from the original cube, which is great. But it doesn't transfer the modifiers. Is there a way to do that?

And the type colors, is there a way to change them? The object black color is really hard to tell agains the dark background, and I'm my 50's, so my eyesight is not the best.

**Reply by OmarEmaraDev** [reply]
@sebarious70 You can check the Copy Full Object option in the Instancer node to copy the modifiers as well.
What do you mean by Type Colors?

**Reply by sebarious70** [reply]
Thanks!

I mean, different classes of inputs and outputs have different colors, like matrices are orange, vectors out is blue, and the object to object is black. Since the background is I'm guessing 90% black, it makes it really hard to see.

**Reply by OmarEmaraDev** [reply]
The colors are not themable unfortunately. But you can change the background color in Blender's theme setting somewhere. Or just choose a different theme that is more contrasty.

**Reply by sebarious70** [reply]
> Maybe look into the Copy Modifiers and Object Material Output nodes to achieved what you want.

Do these would work for lighting up the note that is playing at a given time if I did it with the original setup that doesn't turn each note into a different object? Because I realized that making it that way renders Blender unusable. This is on a Mac Studio M1 Ultra, which is very fast for everything, but having a scene with over 3,000 objects, each with the wireframe modifier, is obviously too much. 

And that's fine, because the original node flow is perfectly fine for what I need, except that I would really like if the cube that represents the note being played at any given time changes the emission from just the edges to the whole cube. But I'm guessing that is not possible unless perhaps I do a very extensive work on the UV map for the mesh that contains the whole group of cubes, and that would take forever for over 3000 notes, am I correct?

**Reply by OmarEmaraDev** [reply]
@sebarious70 You can use vertex colors to change the color of notes in a single mesh. He I basically identify the notes where the current time in seconds in larger than the note start and less than the note end, then I mix colors based on that. Lastly, I add a vertex color layer to the mesh which I can use in the shader. Note that we need to repeat the colors in the list 6 times for each polygon in the cube.

![20240707-113553](https://github.com/JacquesLucke/animation_nodes/assets/25300994/33a13d7b-0950-4ffe-8c13-f29df4238f87)

[midiToMesh.zip](https://github.com/user-attachments/files/16118585/midiToMesh.zip)

However, this is not going to be very efficient either. Maybe an instancing approach would be better here, but I don't have much time to show a solution right now.

**Reply by sebarious70** [reply]
Thanks, I'll give that a try as soon as I get a chance.

Now, you say an instancing approach would be better, do you mean like the second node setup you showed me that turned each note into a separate object? That didn't work very well because it made Blender unresponsive. It was over 3,000 objects I think, so it was too much, even if they were just cubes. I thought maybe one way to make it lighter would be to remove the animation nodes, which is what I had done with the normal setup once it built the meshes, but in this case, once I deleted all the nodes, it deleted all the objects it had created.

**Reply by OmarEmaraDev** [reply]
@sebarious70 I meant using Blender's instancing features, not sure if you ever used those.
About objects getting deleted. You can use the Unlink Instance from Node operator in the Advanced Node Settings in the node options in the N panel. Then you can remove the node tree.

**Reply by sebarious70** [reply]
Oh, no, I haven't used the instancing features in Blender, but I'm sure I can find a YouTube video on them.

Thanks for the tip on the Unlink Instance.

**Reply by sebarious70** [reply]
@OmarEmaraDev Could it be that the "unlink instance from node" operator doesn't exist in the current version? I went up and down all those options several times, and I can't find it. This is what I see where you referred me to:

<img width="505" alt="Screenshot 2024-07-10 at 00 28 33" src="https://github.com/JacquesLucke/animation_nodes/assets/157246475/134c208a-a371-4322-b594-bf86232dfc9b">

As you can see, under Advanced Settings it shows "Make Copy" as a toggle, and then the "Change Type" button, which when pressed shows that list, but I went up and down said list and it doesn't show any unlink.

**Reply by OmarEmaraDev** [reply]
@sebarious70 It seems you don't have the Object Instancer node selected. Those seems like the options for another node.

**Reply by sebarious70** [reply]
@OmarEmaraDev thanks, now I see it. But I also discovered a better way, and I may be able to keep the node tree that separates each note as a different object, because Node Tree settings in the N panel have an Auto Execution check that obviously was punishing Blender because it was trying to build 1576 cubes constantly, while if that's disabled, then it's 1576 objects but they are static. This way it works out better because if I happen to change something, then I wouldn't want to build the whole node tree again. Then if I want to rebuild, I press the "Execute node tree" button.

Although I would have to see if once I add the other tracks that I plan on adding, and this amounts to over 10,000 cubes, if Blender is still going to be responsive having that many objects in the scene. So I guess I'll have to try that.

BTW, is there a way to save a node tree to use in another project? I spent a lot of time working on the other project that had all the notes as one mesh, so I have a lot of animation stuff (not for the animation nodes, but Blender native position and materials, etc) that it would take me too long to bring back into this project, so I would to recreate this node tree in the other project, but I've been wondering if you can just save the node tree to a file, or something like that.

**Reply by OmarEmaraDev** [reply]
@sebarious70 You can just append the node tree to another file. If you wrap your node tree into a subprogram, the subprogram will be accessible in all node trees. Otherwise, use the appended node tree directly.

**Reply by sebarious70** [reply]
@OmarEmaraDev I keep going down the rabbit hole in this project, and I didn't ask you this before because I thought it might be out of scope for AN, but today I was thinking that perhaps it's not. The next step for me would be for the note that is playing currently to be lit up. Let me show you with screenshots. 

Right now I have a wireframe modifier to the Target object, and two materials assigned to it, the first slot is a plain black material, the second one is an emission material, so basically this:

![With modifier copy](https://github.com/user-attachments/assets/c0a3ca61-b531-4bbf-816d-f81fb0771ad7)

I have the camera keyframed to travel from the beginning of the MIDI track to the end. What I want to accomplish is that the note that is playing currently changes from having emission only in the wireframe to have it in its whole surface. To give you an idea, and this is just a mockup:

![Note 1](https://github.com/user-attachments/assets/b9b5abf5-9f4c-4edc-bf02-2a5fd69b3d65)

![Note 2](https://github.com/user-attachments/assets/1196df34-ce13-472f-bab5-93f5a5fbb6cb)

![Note 3](https://github.com/user-attachments/assets/bb7d5cc5-05a8-425d-b8af-babbcc5f8839)

I've been trying to figure that out for several days if not weeks, and no matter how much I search, or read the manual, I can't find the answer. It's easy to get them all lit up or not with geometry nodes:

![Screenshot 2024-07-12 at 18 23 06 copy](https://github.com/user-attachments/assets/49e85745-cfd3-431c-a77a-1b608ec83e7d)

![Screenshot 2024-07-12 at 18 23 18 copy](https://github.com/user-attachments/assets/28cc00c4-f619-489a-9cda-56aa41dcb635)

But, having just one go into full mode sequentially, that's the challenge. So far I had been focusing on the nodes in the shader, then geometry, because I figured AN built the meshes but didn't have much to do with materials.

But then one thing occurred to me; what if AN can assign some type of ID to each cube it duplicates, and that's an ID that somehow can be picked up by geometry nodes so when I change the Set Material Index it only applies to that specific cube, not to all of them. 

Does that sound like something that is possible?

**Reply by OmarEmaraDev** [reply]
What you are describing seems like what I demonstrated in https://github.com/JacquesLucke/animation_nodes/discussions/1962#discussioncomment-9978377 already?
